### ISSUE TITLE
Document fly completion

### ISSUE THREAD

**Issue by jmcarp** [concourse/concourse#1309]
## A Modest Proposal

Fly now supports tab completion for certain flags, but users will have to set up completion in their shells to make use of the feature. For example, my `.zshrc` includes this:

```
#compdef fly
autoload -U compinit && compinit
autoload -U bashcompinit && bashcompinit
_fly_bash_autocomplete() {
    args=("${COMP_WORDS[@]:1:$COMP_CWORD}")
    # Only split on newlines
    local IFS=$'\n'
    # Call completion (note that the first element of COMP_WORDS is
    # the executable itself)
    COMPREPLY=($(GO_FLAGS_COMPLETION=1 ${COMP_WORDS[0]} "${args[@]}"))
    return 0
}
complete -F _fly_bash_autocomplete fly
```

Some cli libraries can auto-generate these invocations for users, like https://github.com/alecthomas/kingpin#bashzsh-shell-completion, but go-flags doesn't. Or if fly were installed with homebrew, we could use a post-install hook to add completion scripts--but we don't. Seems like the options are to add sample bash/zsh invocations to the docs, or possibly update `sync` or another command to output completion scripts from `fly`.

**Comment by suhlig** [comment]
FYI, I just looked at [`_cf-cli`](https://github.com/cloudfoundry/homebrew-tap/blob/master/cf-cli.rb), and simply re-using the same function works fine. I added `complete -F _cf-cli fly` to my `.bashrc`, and now fly has proper command completion.

Maybe fly / Concourse just needs to document this (pretty generic) function somewhere?

**Comment by dohsimpson** [comment]
Thanks to @suhlig 's comment. Since I don't want to install cf-cli. What I did is the following:

1. Create a file like this

```
#!/bin/bash

# credit to: https://github.com/cloudfoundry/homebrew-tap/blob/master/cf-cli.rb
_fly_compl() {
    # All arguments except the first one
    args=("${COMP_WORDS[@]:1:$COMP_CWORD}")
    # Only split on newlines
    local IFS=$'\n'
    # Call completion (note that the first element of COMP_WORDS is
    # the executable itself)
    COMPREPLY=($(GO_FLAGS_COMPLETION=1 ${COMP_WORDS[0]} "${args[@]}"))
    return 0
}

complete -F _fly_compl fly
```

2. source that file in your bashrc or zshrc.

**Comment by mattysweeps** [comment]
A common pattern among tools is to generate this with an `autocomplete` sub command.

ex: `kubectl completion bash` or `helm completion bash`

**Comment by vito** [comment]
@mattysweeps I like that idea! Seems like something easily PR-able so I'll just mark this as 'good first issue/help wanted' since it's unlikely we'll get around to this ourselves.

**Comment by nicoulaj** [comment]
We now have a real completion function for zsh in https://github.com/zsh-users/zsh-completions/commit/b512d57b6d0d2b85368a8068ec1a13288a93d267 !
See zsh-users/zsh-completions#610 for specifics

**Comment by iamjarvo** [comment]
Should this be closed since #4012 was merged?

**Comment by vito** [comment]
Yep, looks like it's documented, too: https://concourse-ci.org/fly.html#fly-completion

thanks!
